components/eamxx/src/physics/mam/eamxx_mam_microphysics_process_interface.cpp

```diff
@@ -89,184 +170,347 @@ set_computed_group_impl(const FieldGroup& group) {
-    2 * aero_config_.num_modes(); // modal number mixing ratio tracers
-  for (int m = 0; m < aero_config_.num_modes(); ++m) {
-    auto m_index = static_cast<mam4::ModeIndex>(m);
-    for (int a = 0; a < aero_config_.num_aerosol_ids(); ++a) {
-      auto a_id = static_cast<mam4::AeroId>(a);
-      if (mam4::aerosol_index_for_mode(m_index, a_id) != -1) {
-        num_aero_tracers += 2; // aerosol mass mixing ratios (interstitial, cloudborne)
-      }
-    }
-  }
-  */
-
-  EKAT_REQUIRE_MSG(group.m_info->size() >= num_aero_tracers,
-    "Error! MAM4 requires at least " << num_aero_tracers << " aerosol tracers.");
+  // how many aerosol/gas tracers do we expect?
+  int num_tracers = 2 * (mam_coupling::num_aero_modes() +
+                         mam_coupling::num_aero_tracers()) +
+                    mam_coupling::num_aero_gases();
+  EKAT_REQUIRE_MSG(group.m_info->size() >= num_tracers,
+    "Error! MAM4 requires at least " << num_tracers << " aerosol tracers.");
 }
 
 size_t MAMMicrophysics::requested_buffer_size_in_bytes() const
 {
-  // number of Reals needed by local views in interface
-  const size_t request = sizeof(Real) *
-    (Buffer::num_2d_mid * ncol_ * nlev_ +
-     Buffer::num_2d_iface * ncol_ * (nlev_+1));
-
-  // FIXME: Need to figure out whether we need this stuff
-  /*
-  // Number of Reals needed by the WorkspaceManager passed to shoc_main
-  const auto policy       = ekat::ExeSpaceUtils<KT::ExeSpace>::get_default_team_policy(m_num_cols, nlev_packs);
-  const int n_wind_slots  = ekat::npack<Spack>(2)*Spack::n;
-  const int n_trac_slots  = ekat::npack<Spack>(m_num_tracers+3)*Spack::n;
-  const size_t wsm_request= WSM::get_total_bytes_needed(nlevi_packs, 13+(n_wind_slots+n_trac_slots), policy);
-  */
-
-  return request;// + wsm_request;
+  return mam_coupling::buffer_size(ncol_, nlev_);
 }
 
 void MAMMicrophysics::init_buffers(const ATMBufferManager &buffer_manager) {
   EKAT_REQUIRE_MSG(buffer_manager.allocated_bytes() >= requested_buffer_size_in_bytes(),
                    "Error! Insufficient buffer size.\n");
 
-  Real* mem = reinterpret_cast<Real*>(buffer_manager.get_memory());
-
-  // set view pointers for midpoint fields
-  using view_2d_t = decltype(buffer_.z_mid);
-  view_2d_t* view_2d_mid_ptrs[Buffer::num_2d_mid] = {
-    &buffer_.z_mid,
-    &buffer_.dz,
-    &buffer_.qv_dry,
-    &buffer_.qc_dry,
-    &buffer_.n_qc_dry,
-    &buffer_.qi_dry,
-    &buffer_.n_qi_dry,
-    &buffer_.w_updraft,
-    &buffer_.q_h2so4_tend,
-    &buffer_.n_aitken_tend,
-    &buffer_.q_aitken_so4_tend,
-  };
-  for (int i = 0; i < Buffer::num_2d_mid; ++i) {
-    *view_2d_mid_ptrs[i] = view_2d_t(mem, ncol_, nlev_);
-    mem += view_2d_mid_ptrs[i]->size();
+  size_t used_mem = mam_coupling::init_buffer(buffer_manager, ncol_, nlev_, buffer_);
+  EKAT_REQUIRE_MSG(used_mem==requested_buffer_size_in_bytes(),
+                   "Error! Used memory != requested memory for MAMMicrophysics.");
+}
+
+namespace {
```

Comment:
1. Totally agree. I just wanted to make clear exactly what is needed here in "elementary" terms, because I'm not super comfortable with the level of abstraction of SCREAM's SCORPIO interface.
2. :+1:. We can even move them to the `impl/` directory.

Reply:
You can move them wherever you want, so long as they are not in the _process interface_ file. But yeah, `impl/` sounds like a proper place.

Reply:
I moved the free functions to `impl/gas_phase_chemistry.cpp`, their rightful home.